rfcs/0000-paths.md

```diff
@@ -0,0 +1,302 @@
+- Feature Name: specify_path_resolution
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [FuelLabs/sway-rfcs#0000](https://github.com/FuelLabs/sway-rfcs/pull/001)
+- Sway Issue: [FueLabs/sway#0000](https://github.com/FuelLabs/sway/issues/001)
+
+# Summary
+
+[summary]: #summary
+
+This RFC proposes a change to the way paths and modules work in Sway.
+
+Paths are sequences of `::`-delimited idents which are used to refer to a
+specific item in a specific module. Paths can either be absolute or relative
+depending on whether they start with a `::` token.
+
+Under this proposal the first identifier in an absolute path is always the name
+of a package. This package either could be a dependency declared in the
+package's Forc.toml file, or the name of the current package. In contrast,
+relative paths are always relative to the current module.
+
+As part of the transition to this new system, this RFC also proposes to replace
+`dep` declarations with `mod` declarations and to deprecate the use of program
+kind headers at the start of Sway source files.
+
+# Motivation
+
+[motivation]: #motivation
+
+There are currently several issues with the way modules and packages are named
+and referred to in Sway.
+
+* The name a library module gives itself does not need to match the name given
+  in the `dep` declaration which includes it.
+* The name the root library module of a library package gives itself does not
+  need to match the name of the package.
+* Due to the above two issues, its possible to have namespace collisions where
+  two separate packages or modules can be imported despite having the same
+  name.
+* It's possible for modules under the root module of package to have name
+  conflicts with dependency packages.
+
+With the new rules proposed in this RFC, all modules that are in scope in (both
+those in the current package and those in dependency packages) have a distinct
+and unique absolute path. What's more, relative paths are always unambiguous
+and refer to a specific item in a specific module.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+## Program kinds in Forc.toml
+
+The `[project]` section of a package's Forc.toml is required to have a `kind`
+field specifying the kind of the program. This value of this field must be one
+of `"library"`, `"contract"`, `"script"` or `"predicate"`. This replaces the
+now-deprecated program kind header at the start of a Sway source file.
```

Comment:
Thought: currently, a project can mix in libraries with a contract/script/predicate. After this change, they cannot. Not inherently a bad thing, especially after workspaces are introduced. Just something to note.